In [0]:
from knn import *
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from scipy.spatial.distance import euclidean
from sklearn.metrics import jaccard_similarity_score as j

**Reads the data set, drops irrelevant variables, and begins to label encode the Y variable**

In [35]:
churn = pd.read_csv('churn_data.csv')
churn = churn.drop(['CustID', 'FamilySize', 'Gender', 'Income'], axis=1)

le = preprocessing.LabelEncoder()
le.fit([0,1])




LabelEncoder()

**Assigns the X variables and Y variables, label encodes the Y variable, and splits the dataset into training and testing data**

In [0]:

x = churn
x = x.drop(['Churn'], axis=1)
y = churn['Churn']
y = le.fit_transform(y)

x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.5, random_state=4)


**Creates the model, fits it, and predicts it**

In [37]:

model = KNN(3, euclidean)

model.fit(x_train, y_train)

results = model.predict(x_test)



(9.486832980505138, 0)
(9.539392014169456, 0)
(39.97499218261337, 0)
(26.70205984563738, 1)
(2.6457513110645907, 1)
(64.89992295835181, 1)
(10.198039027185569, 1)
(18.24828759089466, 1)
(7.280109889280518, 0)
(4.242640687119285, 1)
(26.60826939130014, 1)
(38.58756276314948, 0)
(17.88854381999832, 1)
(48.342527861087284, 0)
(56.753854494650845, 0)
(20.248456731316587, 0)
(9.16515138991168, 0)
(8.660254037844387, 0)
(22.427661492005804, 1)
(22.9128784747792, 1)
(14.491376746189438, 1)
(35.02855977627399, 1)
(12.083045973594572, 1)
(17.69180601295413, 0)
(26.92582403567252, 0)
(58.86425061104575, 1)
(20.12461179749811, 0)
(34.307433596816885, 0)
(19.697715603592208, 1)
(21.93171219946131, 0)
(29.017236257093817, 0)
(5.0, 1)
(29.49576240750525, 0)
(24.020824298928627, 1)
(20.37154878746336, 0)
(33.66006535941367, 1)
(7.0, 0)
(42.14261501141095, 0)
(14.730919862656235, 1)
(20.493901531919196, 1)
(30.166206257996713, 1)
(45.880278987817846, 1)
(40.39801975344831, 0)
(19.87460691435179, 1)
(5

**Shows error metrics**

In [38]:

print('Accuracy: ' + str(accuracy_score(y_test, results)))
print('Precision: ' + str(precision_score(y_test, results)))
print('Recall: ' + str(recall_score(y_test, results)))
print('F1: ' + str(f1_score(y_test, results)))





Accuracy: 0.5625
Precision: 0.5789473684210527
Recall: 0.6470588235294118
F1: 0.6111111111111113


**Looks at the validation dataset**

In [39]:
val = pd.read_csv('churn_validation.csv')

val = val.drop(['CustID', 'FamilySize', 'Gender', 'Income'], axis=1)

le = preprocessing.LabelEncoder()
le.fit([0,1])

LabelEncoder()

In [0]:
x1 = val
x1 = x1.drop(['Churn'], axis=1)
y1 = val['Churn']
y1 = le.fit_transform(y1)

x_train1, x_test1, y_train1, y_test1 = train_test_split(x1,y1, test_size=0.5, random_state=4)

In [41]:
model1 = KNN(3, euclidean)

model1.fit(x_train, y_train)

results1 = model1.predict(x_test)

(9.486832980505138, 0)
(9.539392014169456, 0)
(39.97499218261337, 0)
(26.70205984563738, 1)
(2.6457513110645907, 1)
(64.89992295835181, 1)
(10.198039027185569, 1)
(18.24828759089466, 1)
(7.280109889280518, 0)
(4.242640687119285, 1)
(26.60826939130014, 1)
(38.58756276314948, 0)
(17.88854381999832, 1)
(48.342527861087284, 0)
(56.753854494650845, 0)
(20.248456731316587, 0)
(9.16515138991168, 0)
(8.660254037844387, 0)
(22.427661492005804, 1)
(22.9128784747792, 1)
(14.491376746189438, 1)
(35.02855977627399, 1)
(12.083045973594572, 1)
(17.69180601295413, 0)
(26.92582403567252, 0)
(58.86425061104575, 1)
(20.12461179749811, 0)
(34.307433596816885, 0)
(19.697715603592208, 1)
(21.93171219946131, 0)
(29.017236257093817, 0)
(5.0, 1)
(29.49576240750525, 0)
(24.020824298928627, 1)
(20.37154878746336, 0)
(33.66006535941367, 1)
(7.0, 0)
(42.14261501141095, 0)
(14.730919862656235, 1)
(20.493901531919196, 1)
(30.166206257996713, 1)
(45.880278987817846, 1)
(40.39801975344831, 0)
(19.87460691435179, 1)
(5

In [42]:
print('Accuracy: ' + str(accuracy_score(y_test, results1)))
print('Precision: ' + str(precision_score(y_test, results1)))
print('Recall: ' + str(recall_score(y_test, results1)))
print('F1: ' + str(f1_score(y_test, results1)))


Accuracy: 0.5625
Precision: 0.5789473684210527
Recall: 0.6470588235294118
F1: 0.6111111111111113


**Assignment 3 Write Up**

1) The response variable is whether or not a company has client turn over. It is predicted by age of the client, education of the client, the number of calls to that client, and the visits made to the client. Income, gender, customer ID and family size were omitted as predictor variables. Customer ID obviously does not accurately predict whether or not a client will stay. I did not see the validity of looking at gender either. Family size seemed to also be an irrelevant variable. Income, while having the possibility of being relevant, did not seem a great indicator since not enough was known about the business. We also do not know how elastic the good provided was or the level of available subtitutes. 

2) In order for KNN to work on this data set, the response variable (customer turnover) needed to be label encoded where 0 indicated no customer turnover and 1 indicated customer turnover.


3) KNN was used to build a model to predict customer turnover. Euclidean and Jaccard similarity score were used as different distance metrics. They had the same effect on the model. Different K's were also used but the error metrics stayed the same. Even when adding the seemingly irrelevant variables, the error metrics did not change. 

4) The error metrics used were accuracy, precision,  recall, and F1.. The accuracy metric computes the subset accuracy of the set of labels that exactly matches the corresponding response variables. The precision score is the percent of true positives compared to the number of false positives. Recall then looks at the ratio of true positives divided by the number of true positives and false negatives. F1 is then the weighted average of the precision and recall error metrics.  

5) The model oddly had the same error metrics despite changing the number of predictor variables.The predictors used in the final model were Age, Education, Calls, and Visits. 

6) The error rates I got for the validation dataset where the same as the actual dataset. This must not be a best model for predicting. 

7) Based on the accuracy and precision alone (0.56 and 0.57 respectively), this is not the greatest model for predicting whether or not a customer will leave a company. The lack of variation across specifications leaves me to question the validity of the model.